search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,498 @@
+    }
+
+    if (self->skeleton != NULL) {
+        g_dbus_interface_skeleton_unexport (G_DBUS_INTERFACE_SKELETON (self->skeleton));
+        g_clear_object (&self->skeleton);
+    }
+
+    g_clear_object (&self->object_manager);
+    g_clear_object (&self->doc);
+    g_clear_pointer (&self->page_data_hash_map, g_hash_table_unref);
+
+    G_OBJECT_CLASS (yelp_search_provider_app_parent_class)->dispose (obj);
+}
+
+static void
+preload_data_cb (YelpDocument          *document,
+                       YelpDocumentSignal    signal,
+                       YelpSearchProviderApp *app,
+                       GError                *error)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (app);
+    gchar **page_ids = yelp_document_list_page_ids (document);
+    gchar **iter;
+
+    for (iter = page_ids; *iter; iter++){
+        gchar *page_id = *iter;
+        PageData *data = g_new0 (PageData, 1);
+        gchar *icon_string = yelp_document_get_page_icon (document, page_id);
+        GIcon *icon = g_icon_new_for_string (icon_string, &error);
+        g_free (icon_string);
+
+        data = page_data_new_steal (yelp_document_get_page_title (document, page_id), 
+                                yelp_document_get_page_desc (document, page_id), 
+                                icon);
+        g_hash_table_insert (app->page_data_hash_map, page_id, data);
+    }
+
+    self->name_owner_id = g_bus_own_name (G_BUS_TYPE_SESSION,
+                                         "org.gnome.Yelp.SearchProvider",
+                                         G_BUS_NAME_OWNER_FLAGS_NONE,
+                                         search_provider_bus_acquired_cb,
+                                         search_provider_name_acquired_cb,
+                                         search_provider_name_lost_cb,
+                                         app, NULL);
+    g_free (page_ids);
+    app->doc = g_object_ref (document);
+}
+
+static void
+uri_resolved_cb (YelpUri  *uri,
+                 YelpSearchProviderApp *app)
+{
+    YelpDocument *doc;
+    gchar *page_id;
+
+    doc = yelp_document_get_for_uri (uri);
+    g_signal_handler_disconnect (uri, app->id);
+    page_id = yelp_uri_get_page_id (uri);
+    yelp_document_request_page (doc,
+                               page_id,
+                               NULL,
+                               (YelpDocumentCallback) preload_data_cb,
+                               app,
+                               NULL);
+
+    g_object_unref (uri);
+    g_object_unref (doc);
+}
+
+static void
+search_provider_app_startup (GApplication *app)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (app);
+    YelpUri *base_uri;
+    G_APPLICATION_CLASS (yelp_search_provider_app_parent_class)->startup (app);
+
+    // TODO: remove holding the app forever
+    if (g_getenv ("YELP_SEARCH_PROVIDER_PERSIST") != NULL)
+        g_application_hold (app);
+
+    base_uri = yelp_uri_new ("help:gnome-help");
```

Comment:
I think you should unref the `base_uri` after calling `yelp_uri_resolve` at the end of the function - `yelp_uri_resolve` will take a ref on this variable on its own.